modules/sdk/src/test/scala/org/tessellation/sdk/infrastructure/collateral/daemon/CollateralDaemonSuite.scala

```diff
@@ -0,0 +1,109 @@
+    SecurityProvider.forAsync[IO]
+
+  private val (address1, peer1) = (
+    Address("DAG0y4eLqhhXUafeE3mgBstezPTnr8L3tZjAtMWB"),
+    PeerId(
+      Hex(
+        "6128e64d623ce4320c9523dc6d64d7d93647e40fb44c77d70bcb34dc4042e63cde16320f336c9c0011315aa9f006ad2941b9a92102a055e1bcc5a66ef8b612ef"
+      )
+    )
+  )
+
+  private val (address2, peer2) = (
+    Address("DAG07tqNLYW8jHU9emXcRTT3CfgCUoumwcLghopd"),
+    PeerId(
+      Hex(
+        "79c4a78387a8782dbc88de95098d134a7dbf3b8a3316eaa1e41e112dc5b21a5b0cefdd0871495435591089264aa5c8a2429a75b384519662184bedfa6e7b886f"
+      )
+    )
+  )
+
+  def mkLatestBalances(balancesRef: SignallingRef[IO, Option[Map[Address, Balance]]]) =
+    new LatestBalances[IO] {
+      def getLatestBalances: F[Option[Map[Address, Balance]]] = balancesRef.get
+
+      def getLatestBalancesStream: Stream[F, Map[Address, Balance]] =
+        balancesRef.discrete.flatMap(_.fold[Stream[F, Map[Address, Balance]]](Stream.empty)(Stream(_)))
+    }
+
+  def mkClusterStorage() = {
+    val peers = Map(peer1 -> peerGen.sample.get.copy(id = peer1), peer2 -> peerGen.sample.get.copy(id = peer2))
+    ClusterStorage.make[IO](ClusterId("8d07c061-d42f-4d9c-9efc-37e0d1ee73e7"), peers)
+  }
+
+  def mkCollateralDaemon(latestBalances: LatestBalances[IO], clusterStorage: ClusterStorage[IO])(
+    implicit sc: SecurityProvider[IO]
+  ) = {
+    val collateralConfig = CollateralConfig(Amount(25_000_000_000_000L))
+    val collateral = Collateral.make[IO](collateralConfig, latestBalances)
+    CollateralDaemon.make(collateral, latestBalances, clusterStorage)
+  }
+
+  test("peers stay in cluster when collateral is sufficient") { implicit sc =>
+    for {
+      balancesRef <- SignallingRef.of[IO, Option[Map[Address, Balance]]](
+        Some(Map(address1 -> Balance(25_000_000_000_000L), address2 -> Balance(25_000_000_000_000L)))
+      )
+      latestBalances = mkLatestBalances(balancesRef)
+      clusterStorage <- mkClusterStorage()
+      collateralDaemon = mkCollateralDaemon(latestBalances, clusterStorage)
+      control <- TestControl.execute(
+        collateralDaemon.start >> balancesRef.set(
+          Map(address1 -> Balance(25_000_000_000_000L), address2 -> Balance(25_000_000_000_000L)).some
+        )
+      )
+      _ <- control.tickAll
+      peers <- clusterStorage.getPeers
+      peersIds = peers.map(_.id)
+    } yield expect.same(peersIds, Set(peer1, peer2))
+  }
+
+  test("peers are removed from cluster when collateral is not sufficient") { implicit sc =>
+    for {
+      balancesRef <- SignallingRef.of[IO, Option[Map[Address, Balance]]](
+        Some(Map(address1 -> Balance(25_000_000_000_000L), address2 -> Balance(25_000_000_000_000L)))
+      )
+      latestBalances = mkLatestBalances(balancesRef)
+      clusterStorage <- mkClusterStorage()
+      collateralDaemon = mkCollateralDaemon(latestBalances, clusterStorage)
+      control <- TestControl.execute(
```

Comment:
Why you need a test control for that? Can't you just execute services directly?

Reply:
Without `TestControl` peers returned by `getPeers` method are not updated yet. Because there is used `Spawn` under `start` method it looks like update of `clusterStorage` is scheduled after `getPeers` is called in test.

Reply:
Okay, got it